Code compliant chiller performance curves

lymereJ opened the pull request:
Pull request overview
---------------------
This PR adds a new dataset, `CodeCompliantEquipment.idf`, that includes chiller performance curves to be use to meet the minimum requirements in ASHRAE 90.1-2019 and its Appendix G. We propose that the same dataset be improved and supplemented at a later date with additional performance curves for other equipment types (for instance: IEER-rated DX equipment). A NFP was included [here](https://github.com/NREL/EnergyPlus/pull/9046/files#diff-0cb9fc659ed36df5cb2af1fe1b63885c2198f5fdc9a364a40bd5eb8995b542a0) to document, at a high level, the approach used to generate the curves.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

lymereJ committed:
Add code compliant equipment dataset.

lymereJ committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into develop

Myoldmopar commented:
I don't have any concern with this going in, but it needs to be added to the data set test so that we don't accidentally change the Curve inputs and forget to update this file.  https://github.com/NREL/EnergyPlus/blob/develop/tst/EnergyPlus/unit/Datasets.unit.cc

lymereJ committed:
Add new unit test for the dataset.

lymereJ committed:
Merge branch 'develop' into code_comp_eqp

lymereJ commented:
> I don't have any concern with this going in, but it needs to be added to the data set test so that we don't accidentally change the Curve inputs and forget to update this file. https://github.com/NREL/EnergyPlus/blob/develop/tst/EnergyPlus/unit/Datasets.unit.cc

Ah yes, I overlooked it, sorry. It is corrected now.

Myoldmopar commented:
Thanks for adding the unit test.  I personally don't have any issue with this, but I'm wondering if anyone else has any questions or interest.  Since it is equipment curves and code compliance, I'll tag a few relevant folks but by all means anyone who wants to comment, go ahead before this goes in.  

@JasonGlazer @mjwitte @rraustad @EnergyArchmage @lgu1234

rraustad commented on design/FY2021/NFP-CodeCompliantEquipmentDataset.md:
```diff
@@ -0,0 +1,65 @@
+
+## Approach ##
+Generating chiller performances curves meeting both full and part load efficiency numbers require solving a largely underdetermined system of equations. The number of unknowns depend on the type of curves, which mostly depend on the algorithm used by the software and used in models. It also means that an infinite number of solutions exists. One way to solve this issue is to turn it into a constra ...
+
```
Did you mean "generic"

... constrained problem and use a simple genetic algorithm (GA) ...

rraustad commented on design/FY2021/NFP-CodeCompliantEquipmentDataset.md:
```diff
@@ -0,0 +1,65 @@
+Generating chiller performances curves meeting both full and part load efficiency numbers require solving a largely underdetermined system of equations. The number of unknowns depend on the type of curves, which mostly depend on the algorithm used by the software and used in models. It also means that an infinite number of solutions exists. One way to solve this issue is to turn it into a constra ...
+
+The implemented algorithm, which uses a GA, starts with a “seed” set of curves that can be user-modifiable, target full load, and part load efficiency as inputs was designed. The GA reproduces the process of natural selection on an initial population and subsequent generations. Individuals in the initial population are generated from the seed curves by randomly modifying the coefficient of each c ...
+
```
And if there are edits there is an extra space here.

... far their results  are from the target.

lymereJ replied:
Thanks! I removed the extra space.

mjwitte commented on datasets/CodeCompliantEquipment.idf:
```diff
@@ -0,0 +1,2427 @@
+!- Beginning of performance curves for Chiller:Electric:EIR:
+
+Curve:Quadratic,
+    ASHRAE901_AppG_wtr_cent_gte150lt300ton_0.634kWpton_0.596IPLV.IP_eir-f-plr,    !- Name
```
Not a huge problem, but it appears that all of the eir-f-plr curves are at the top, then the other curves are grouped by chiller size etc.

JasonGlazer commented:
@lymereJ this is a wonderful addition to the EnergyPlus datasets. My only comment is about the documentation.  Maybe copy the text from the NFP into the normal EnergyPlus documentation that explains how it was made as part of the Output Details and Examples file. 

Also, it would be great at the top of the CodeCompliantEquipment.idf file to add some comments about how to parse all the pieces that make up the names of each of the curves. The name "ASHRAE901_2019_wtr_A_pdsp_gte264lt528kW_4.889COP_6.286IPLV.SI_eir-f-plr" has a lot of information in it but many people may not recognize what the "pdsp" means or even the "A."

Finally, at the top of the CodeCompliantEquipment.idf file please document how the PLR and temperature curves should be used together. I presume the names are related except for the _plr and _t but I'm not sure.

mjwitte commented on datasets/CodeCompliantEquipment.idf:
```diff
@@ -0,0 +1,2427 @@
+    0.03157995838320167,      !- Coefficient2 x
+    0.6470161436053915,       !- Coefficient3 x2
+    0.0,                      !- Minimum Value of x
+    1.0,                      !- Maximum Value of x
```
This means the input PLR will be capped at 1.0, even if the chiller has a Maximum Part Load Ratio > 1.0. Is that ok here?

lymereJ replied:
That's our recommendation, and because the curves aren't normative, users could overwrite to extrapolate if need be.

rraustad commented on design/FY2021/NFP-CodeCompliantEquipmentDataset.md:
```diff
@@ -0,0 +1,65 @@
+N/A
+
+## Example File and Transition Changes ##
+N/A
```
Shoot, I got down here and realized I was reading the NFP.

mjwitte commented on datasets/CodeCompliantEquipment.idf:
```diff
@@ -0,0 +1,2427 @@
+    -0.013818136673476294,    !- Coefficient4 y
+    0.0006361097011564334,    !- Coefficient5 y2
+    -0.0002740904902173277,    !- Coefficient6 xy
+    4.0,                      !- Minimum Value of x
+    15.6,                     !- Maximum Value of x
+    10.0,                     !- Minimum Value of y
+    40.0,                     !- Maximum Value of y
```
Likewise, these inputs will be constrained even if the loop temps or setpoints go outside this range. Chilled water temps 4-15.6C (39.2-60.08 F) and condenser water temps of 10-40C (50-104 F). Are these ranges adequate?

lymereJ replied:
We believe so. We could re-generate them with extended temperature ranges if need be.

mjwitte reviewed (commented):
This is a very useful dataset. As described in the NFP, it's not trivial to come up with these.

Because Chiller:Electric:EIR has flexible inputs for Reference Leaving Chilled Water Temperature and Reference Entering Condenser Fluid Temperature, it would be good to document here what was assumed for these curves. Are the fT curves all near 1.0 at the default reference temps of 6.67 and 29.4C?

rraustad commented:
I agree with the comments thus far but the importance of this data set is obvious. It would have been nice to retain the methodology in the documentation, however, none of the other data sets have a similar description. Same for the references. A data set document could always, however, deviate from past documents. The organization of curves may also be an issue for some users, but the data is assumed to be present. Regarding the curve limits, it is standard practice to use the same limits of the data set, however, when operating outside those limits the results are inaccurate. That is a standing issue with all performance curves. So will leave it to the developer to take or reject any of these comments.

Myoldmopar commented:
I'll wait for @lymereJ to reply on here, but I think this is probably close or all ready to go in once we wrap up discussion.

lymereJ commented:
Thanks everyone for the review and comments! I agree with pretty much everything that was said. I was waiting to see if other comments might be coming in before doing anything. I'll address them in the morning.

rraustad commented:
@lymereJ I think the biggest thing in my mind is ownership of this data set. I also think most users will understand where they come from but a reference to a future paper or something like that would be nice. Also, these curves should be used in the baseline model and may or may not be used in the proposed. It would be nice to also document that usage aspect.

lymereJ committed:
Address comments.

lymereJ committed:
Merge branch 'develop' into code_comp_eqp

lymereJ commented:
@Myoldmopar, @JasonGlazer, @mjwitte, @rraustad  - I _think_ that I've addressed all your comments with the last commit.
* Added text to the documentation (@rraustad we are planning on writing a paper, we'll update that section once and if it gets published)
* Grouped the curves by requirement
* Added comments to describe how the name of each curve is put together
* Added comments about reference temperatures
* Added a comment about authorship

Please, do let me know if I have missed anything or if you believe that something should be revised.

lymereJ committed:
Fix typo [decent_ci_skip].

lymereJ committed:
Fix typo #2 [decent_ci_skip].

mjwitte reviewed (approved):
@lymereJ Thanks for addressing the comments, I'm good with this.

EnergyArchmage commented on doc/output-details-and-examples/src/data-sets/simple-list-data-sets.tex:
```diff
@@ -66,6 +66,31 @@ \subsection{Chillers.idf}\label{chillers.idf}
+
+This dataset includes performance curve objects that describe the part load performance of building energy code minimally compliant equipment. The dataset provides curves for the following EnergyPlus objects: Chiller:Electric:EIR. Sets of curves are provided for the following building energy codes minimum requirements: ASHRAE Standard 90.1-2019 Table 6.8.1-3, ASHRAE Standard 90.1-2019 Table G3.5. ...
+
+Generating performances curves meeting both full and part load efficiency numbers require solving a largely underdetermined system of equations. The number of unknowns depend on the type of curves, which mostly depend on the algorithm used by the software and used in models. It also means that an infinite number of solutions exists. One way to solve this issue is to turn it into a constrained pro ...
```
Would it be possible to describe some specifics for how the cost function was defined?

Myoldmopar replied:
@lymereJ can you respond to @EnergyArchmage here in this PR and then sometime after release add an entry to the documentation or dataset header or something that also captures this?

lymereJ replied:
@EnergyArchmage @Myoldmopar -

Sorry for the belated response. Here's the fitness function that we used, presented in "pseudocode":

`fitness = (iplv_score + full_load_score + normalization_score + 0.5 * rsme_score) / 3.5`

Where:
- `iplv_score` is the absolute difference between the calculated IPLV and the target
- `full_load_score` is the absolute difference between the calculated full load efficiency and the target
- `normalization_score` is a calculated value that expresses how close to 1 each curve are at rated conditions
- `rmse_score` RMSE between the curves and "seed" curves for a range of data points, the idea here is to encourage curves being "closer" to the seed curves

Note: all score are arbitrarily weighted, all of them are equally weighted except the `rsme_score`.

EnergyArchmage replied:
Thanks for this,  but what is the "target"?  MFG reported test results for the IPLV points?  I can wait for the paper.  Anyway this is great stuff.

Myoldmopar commented:
I'm going to merge this, but @lymereJ please still respond to @EnergyArchmage 's request.
